### ISSUE TITLE
SR failure when permuting two arguments

### ISSUE THREAD

**Issue by francoisthire** [rlepigre/lambdapi#61]
This example comes from this [issue](https://github.com/Deducteam/Dedukti/issues/72).

```
A : Type.

a : A.

def T : A -> Type.

[] T a --> A -> A.

True : Type.

def eq : A -> A -> Type.
[x] eq x x --> True.
true : True.
def f : x : A -> y : A -> eq x y -> T x ->  A.

(; OK ;)
[] f a _ true (x => x) --> a.

(; Not OK ;)
[] f _ a true (x => x) --> a.
```

**Comment by fblanqui** [comment]
Next time, thanks @francoisthire for adding the error message in your issue description. ;-)

[zo.dk, 20:0-28] Error on command.
Product expected, [T ?#1] given.

**Comment by fblanqui** [comment]
As already mentioned for Dedukti, what ever is the structure of the LHS, if it is typable, then it is of type A as the RHS. So these rules preserve typing trivially. It is therefore annoying that Lambdapi fails on the second rule for f.

When trying to type the LHS of the first rule of f, we get the constraints True = eq a y and X -> X = T a. Since T a --> A -> A, we get X=A.

On the other hand, when trying to type the LHS of the second rule of f, we get the constraints True = eq x a and X -> X = T x. Here, Lambdapi fails because T x is not a product. This is a bug. Instead, it should keep this constraint as unsolved.

If the type checking of a LHS fails because there is a truly unsolvable constraint, SR is trivially satisfied but a warning should be issued because the rule can never be applied.

Now, it is undecidable to say whether some constraint is satisfiable or not, for instance to known whether a LHS has a typable instance. Here, it relies on the fact that (eq a y) reduces to True ONLY IF y=a. This is non trivial. It requires to implement narrowing.

**Comment by fblanqui** [comment]
This would require to fix the following two lines:
https://github.com/rlepigre/lambdapi/blob/8d249f47d9350ef99a1fbc10f6034b03568bb6cf/src/typing.ml#L97
https://github.com/rlepigre/lambdapi/blob/8d249f47d9350ef99a1fbc10f6034b03568bb6cf/src/typing.ml#L141
We could add a new case, namely when c is headed by a defined symbol that could reduce to a product. In this case, we would do like in the Meta case: add a constraint. But when is it correct to do this? And does it preserve termination of the algorithm?

**Comment by rlepigre** [comment]
After some consideration, I think that it is better to only add a constraint, regardless of the shape of the term that is expected to be a product. This is what I implemented in 1a33e8fcfdcf5a2c42b77e973d8aa05572d27dad.

The example above is now accepted.

@fblanqui: Should we close or is there something else to discuss?
